src/library/scala/collection/immutable/NVector.scala

```diff
@@ -0,0 +1,2010 @@
+    if(k == 0) this
+    else appendedAll0(suffix, k)
+  }
+
+  protected[this] def appendedAll0[B >: A](suffix: collection.IterableOnce[B], k: Int): NVector[B] = {
+    if(k > 0 && k < tinyAppendLimit) {
+      var v: NVector[B] = this
+      suffix match {
+        case it: Iterable[_] => it.foreach(x => v = v.appended(x))
+        case _ => suffix.iterator.foreach(x => v = v.appended(x))
+      }
+      v
+    } else new NVectorBuilder[B].initFrom(this).addAll(suffix).result()
+  }
+
+  override def className = "NVector"
+
+  @inline override final def take(n: Int): NVector[A] = slice(0, n)
+  @inline override final def drop(n: Int): NVector[A] = slice(n, length)
+  @inline override final def takeRight(n: Int): NVector[A] = slice(length - mmax(n, 0), length)
+  @inline override final def dropRight(n: Int): NVector[A] = slice(0, length - mmax(n, 0))
+  override def tail: NVector[A] = slice(1, length)
+  override def init: NVector[A] = slice(0, length-1)
+
+  protected[this] def slice0(lo: Int, hi: Int): NVector[A]
+
+  override final def slice(from: Int, until: Int): NVector[A] = {
+    val lo = mmax(from, 0)
+    val hi = mmin(until, length)
+    val newlen = hi - lo
+    if(newlen == length) this
+    else if(newlen <= 0) NVector0
+    else slice0(lo, hi)
+  }
+
+  /** Number of slices */
+  protected[immutable] def vectorSliceCount: Int
+  /** Slice at index */
+  protected[immutable] def vectorSlice(idx: Int): Array[_ <: AnyRef]
+  /** Dimension of the slice at index */
+  protected[immutable] final def vectorSliceDim(idx: Int): Int = {
+    val c = vectorSliceCount/2
+    c+1-abs(idx-c)
+  }
+  /** Length of all slices up to and including index */
+  protected[immutable] def vectorSlicePrefixLength(idx: Int): Int
+
+  override def iterator: Iterator[A] = new NVectorIterator(this)
+
+  override def copyToArray[B >: A](xs: Array[B], start: Int, len: Int): Int = iterator.copyToArray(xs, start, len)
+
+  //override def toVector: Vector[A] = this
+
+  //override protected def applyPreferredMaxLength: Int = Vector.defaultApplyPreferredMaxLength
+
+  override def stepper[S <: Stepper[_]](implicit shape: StepperShape[A, S]): S with EfficientSplit = {
+    import convert.impl._
+    val s = shape.shape match {
+      case StepperShape.IntShape    => new IntNVectorStepper(new NVectorIterator(this).asInstanceOf[NVectorIterator[Int]])
+      case StepperShape.LongShape   => new LongNVectorStepper(new NVectorIterator(this).asInstanceOf[NVectorIterator[Long]])
+      case StepperShape.DoubleShape => new DoubleNVectorStepper(new NVectorIterator(this).asInstanceOf[NVectorIterator[Double]])
+      case _                        => shape.parUnbox(new AnyNVectorStepper[A](new NVectorIterator(this)))
+    }
+    s.asInstanceOf[S with EfficientSplit]
+  }
+}
+
+
+/** Empty vector */
+private final object NVector0 extends NVector[Nothing] {
+  import NVectorStatics._
+
+  def length = 0
+  def apply(index: Int) = throw new IndexOutOfBoundsException
```

Comment:
The old vector gives a more detailed message in its `IIOBE` including `i` and the size of this collection, we should keep that behaviour.